docs/design/metainfo-rpc.md

```diff
@@ -0,0 +1,287 @@
+# Metainfo RPC Service Refactor
+
+## Background
+
+Our current metainfo rpc service isn't very future proof.
+If we shipped our current client code, it wouldn't be easy to update clients and add support for the following:
+- versions
+- multipart uploads
+- cleanup for uncommitted segments
+- general concurrency fixes
+
+Native multipart upload can't happen with our current architecture because there are issues with out of order uploading of a single stream.
+
+Currently, concurrent uploading causes segments to get mangled, but adding versions is a good way to handle concurrent uploads.
+The current protobuf does not contain which version is being written, so it has to assume "version 0", which is problematic with concurrent clients (they both are writing to the "version 0").
+
+This should also reduce the number of roundtrips needed to start and end both streams and segments.
+
+This design should ensure compatibility between client and server in the future as these features are implemented. Neither clients nor servers will be able to access more information than what we provide in this new proto, so we need to make sure the messages and service calls anticipate needs for the above features.
+
+## Design
+
+The following is a proto file for just the rpc portion to create segments. The read portions just need to have a stream id included, which should default to being the largest stream id.
+(When we implement versions, every upload will have a new stream, which will be given the largest stream id at that time. This will allow us to get older versions by checking the stream id.)
+
+With this design, streams will be created that contain multiple parts. There were already RPCs to finalize individual segments, but each segment was not logically associated with a part, but instead just the specific object.
+
+```protobuf
+service Metainfo {
+    rpc CreateBucket(BucketCreateRequest) returns (BucketCreateResponse);
+    rpc GetBucket(BucketGetRequest) returns (BucketGetResponse);
+    rpc DeleteBucket(BucketDeleteRequest) returns (BucketDeleteResponse);
+    rpc ListBuckets(BucketListRequest) returns (BucketListResponse);
+
+    rpc CreateObject(ObjectCreateRequest) returns (ObjectCreateResponse);
+    rpc CommitObject(ObjectCommitRequest) returns (ObjectCommitResponse);
+    rpc ListObjects(ObjectListRequest) returns (ObjectListResponse);
+
+    rpc DeleteObject(ObjectDeleteRequest) returns (ObjectDeleteResponse);
```

Comment:
@zeebo  should this call be `BeginDeleteObject`? Also, should this call and `FinishDeleteObject` share `ObjectDeleteRequest` or should there be an `ObjectBeginDeleteRequest` and `ObjectFinishDeleteRequest`?

Reply:
I went ahead and added both `ObjectBeginDeleteRequest` and `ObjectFinishDeleteRequest`